### PR TITLE
Lookup

### PR DESCRIPTION
**Pull request by mvcisback**

### CODE REVIEW

File: aigerbv/aigbv.py [hunk 1]
```diff
@@ -122,15 +122,21 @@ def feedback(self, inputs, outputs, initials=None, latches=None,
 
         idrop, imap = fn.lsplit(lambda x: x[0] in inputs, self.input_map)
         odrop, omap = fn.lsplit(lambda x: x[0] in outputs, self.output_map)
-        new_latches = [(n, common.named_indexes(n)) for n in latches]
+
+        wordlens = [len(vals) for _, vals in self.input_map]
+        new_latches = [(n, common.named_indexes(k, n))
+                       for k, n in zip(wordlens, latches)]
+
+        if initials is not None:
```
Comment (MarkusRabe) [inline aigerbv/aigbv.py:130]: How about assert len(initials) == len(inputs)

Same for latches above.

File: aigerbv/common.py [hunk 2]
```diff
@@ -81,6 +80,22 @@ def bitwise_negate(wordlen, input='x', output='not x'):
...
+    queue = [identity_gate(wordlen, i) for i in inputs]
+    while len(queue) > 1:
+        queue = list(starmap(join, zip(queue[::2], queue[1::2])))
+
+    circ = queue[0]
+    if len(inputs) & 1:
+        circ = join(circ, identity_gate(wordlen, inputs[-1]))
+    return circ
+
+
```
Comment (MarkusRabe) [inline aigerbv/common.py]: Nice, but looks a little complicated. Took me a while to understand line 91. Maybe decompress?

Can line 93 to 95 be merged into the first loop? Not sure what's going on.

File: aigerbv/common.py [hunk 3]
```diff
@@ -401,11 +419,21 @@ def abs_gate(wordlen, input, output):
         >> bitwise_xor(wordlen, output, tmp2, output)
 
 
-def lookup(wordlen, mapping, input, output):
+def lookup(inlen, outlen, mapping, input, output, *,
+           in_signed=True, out_signed=True):
```
Comment (MarkusRabe) [inline aigerbv/common.py:427]: Maybe add an example as a comment how to use this function. The function is a bit too complicated to figure out what the star does.

**Review by MarkusRabe — COMMENTED** [review]
See minor comments

File: aigerbv/aigbv.py (hunk 1 above)
Comment (mvcisback) [reply aigerbv/aigbv.py:130]: Done.

File: aigerbv/common.py (hunk 2 above)
Comment (mvcisback) [reply aigerbv/common.py]: The last part is handling even vs odd case. I'm building the circuit as a balanced tree, by pairing every element. That's what line 91 does in a bit a convoluted way. What I will do is make a helper function called `_apply_pairwise` that will hopefully make it more readable.

File: aigerbv/common.py (hunk 3 above)
Comment (mvcisback) [reply aigerbv/common.py:427]: `*` means keyword only argument in python 3. It's useful for adding features later. I particularly like it for cases where I think not using keywords makes things confusing.

Comment (mvcisback) [reply aigerbv/common.py:427]: That said, I agree, I should document this function better.

**Pull request merged**